docs/proposals/improve-status-in-CAPI-resources.md

```diff
@@ -0,0 +1,921 @@
+    - [Changes to KubeadmControlPlane (KCP) resource](#changes-to-kubeadmcontrolplane-kcp-resource)
+      - [KubeadmControlPlane Status](#kubeadmcontrolplane-status)
+      - [KubeadmControlPlane (New)Conditions](#kubeadmcontrolplane-newconditions)
+      - [KubeadmControlPlane Print columns](#kubeadmcontrolplane-print-columns)
+    - [Changes to MachinePool resource](#changes-to-machinepool-resource)
+    - [Changes to Cluster API contract](#changes-to-cluster-api-contract)
+
+# Summary
+
+This documents defines how status in CAPI resources is going to evolve in the v1beta2 API version, with the goal of
+improving usability and consistency across different resources in CAPI and with the rest of the ecosystem.
+
+# Motivation
+
+The Cluster API community recognize that nowadays Cluster API and Kubernetes users are rightfully focused on
+building higher systems and great applications on top on those platforms, which is great.
+
+However, as a consequence of this shifted focus, most of the users don’t have time to become deep expert of Cluster API
+like the first wave of adopters, and also Cluster API maintainers would like they don’t have to.
+
+The effect of the trend above is the blurring of the lines not only between different Cluster API components, but also
+between Cluster API, core Kubernetes and a few other broadly adopted tools like Helm or Flux (and to some extents,
+also with many others awesome tools in the ecosystem).
+
+This is why Cluster API status must become simpler to understand for users, and also more consistent not only across
+different CAPI resources, but with Kubernetes core and ideally with the entire ecosystem.
+
+### Goals
+
+- Review and standardize the usage of the concept of readiness across Cluster API resources.
+    - Drop or amend improper usage of readiness
+    - Make the concept of Machine readiness extensible, thus allowing providers or external system to inject their readiness checks.
+- Review and standardize the usage of the concept of availability across Cluster API resources.
+    - Make the concept of Cluster Availability extensible, thus allowing providers or external system to inject their availability checks.
+- Bubble up more information about both CP and worker Machines, ensuring consistent way across Cluster API resources.
+    - Standardize replica counters and bubble them up to the Cluster resource.
+    - Standardize control plane, MachineDeployment, Machine pool availability, and bubble them up to the Cluster resource.
+- Introduce missing signals about connectivity to workload clusters, thus enabling to mark all the conditions
+  depending on such connectivity being working with status Unknown after a certain amount of time.
+- Introduce a cleaner signal about Cluster API resources lifecycle transitions, e.g. scaling up or updating.
+- Ensure everything in status can be used as a signal informing monitoring tools/automation on top of Cluster 
```

Comment:
This is great goal. As a service provider I want to be able to use conditions as SLIs to measure that my SLOs are met.

Reply:
Agree. One easy path to that should be: conditions => state metrics => some simple dashboards / alerting rules

Reply:
Thanks for the feedback, resolving to keep the number of active threads under control 😅